workers/approov-token-binding/jwt-verifier.js

```diff
@@ -0,0 +1,127 @@
+
+    // Token failed validation or is missing in the request
+    return new Response(jsonErrorResponse(error, event.request, token_payload), { headers: { 'content-type': 'application/json' }, status: 401 })
+  }
+}
+
+function ApproovTokenBindingException(message) {
+   this.message = message;
+   this.name = 'ApproovTokenBindingException';
+}
+
+function ApproovTokenBindingResult(message) {
+   this.message = message;
+   this.name = 'ApproovTokenBindingResult';
+}
+
+function handleApproovTokenBindingCheck(request, token_payload) {
+
+  let token_binding_header
+  const token_binding = token_payload.pay
+
+  if (token_binding === undefined) {
+    return new ApproovTokenBindingResult("Approov token payload is missing the key 'pay'.")
+  }
+
+  if (isEmptyString(token_binding)) {
+    return new ApproovTokenBindingResult("Approov token payload key 'pay' is empty.")
+  }
+
+  // We default to the Authorization token, but feel free to use another header
+  // and bind it to the Approov token in the mobile app.
+  if (typeof TOKEN_BINDING_HEADER_NAME === 'undefined') {
+    token_binding_header = request.headers.get('Authorization')
+  } else {
+    token_binding_header = request.headers.get(TOKEN_BINDING_HEADER_NAME)
+  }
+
+  if (isEmptyString(token_binding_header)) {
+    throw new ApproovTokenBindingException("Missing or empty header to perform the verification for the Approov token binding.")
+  }
+
+  // We need to hash and base64 encode the token binding header, because that's how it was included in the Approov
+  // token on the mobile app.
+  const token_binding_header_encoded = crypto.createHash('sha256').update(token_binding_header, 'utf-8').digest('base64')
+
+  if (token_binding !== token_binding_header_encoded) {
+    throw new ApproovTokenBindingException("Approov token binding not matching.")
+  }
+
+  return "Aprroov token binding matches."
+}
+
+function log(result, request, token) {
+  const ip = request.headers.get('cf-connecting-ip') || 'Unknown';
+  const userAgent = request.headers.get('User-Agent') || 'Unknown';
+  const requestMethod = request.method || 'Unknown';
+  const requestUrl = request.url || 'Unknown';
+  const safeToken = token || 'Unknown';
+  const result_message = result.message || 'Unknown'
+  const message = "[" + ip + "] " + requestMethod + " " + requestUrl + " [UA: " + userAgent + "] [JWT: " + safeToken + "] [RESULT: " + result_message + "]"
+
+  // If you want access to the logs you need to use the command `wrangler tail`
+  // to see them in real-time from your computer or send you need to add code in
+  // here to send the log message to your preferred cloud based logging
+  // framework.
+  //
+  // Check the `Useful Links` section in the README for more info on the logging
+  // approaches.
+  console.log(message)
+}
+
+function jsonErrorResponse(error, request, token_payload) {
+  let ALLOWED_IPS = []
+
+  if (typeof IP_DEBUG_WHITELIST !== 'undefined') {
```

Comment:
Do we need this - shouldn't we point to whitelist security policies instead

Reply:
Cloudflare is a pain to debug, this will allow to send extra debug information in a response for a failed Approov validation, but only to the allowed IPs on that list. Bear in mind that failures may mean no Approov token at all or an attempt to fake one, therefore we cannot use the Approov security policies.

This IP white list is set as an environment variable in Cloudflare, via the wrangler CLI tool or via the web interface, therefore IPs can be added/removed at anytime without the need to redeploy the Approov Cloudflare worker, and this can be used in any environment, production, staging, testing or development.

Reply:
Okay leave as is then